File: src/data/types/assignShipRoleResult.ts
import type { CrewMember } from "./crewMember.ts";

// Ship Crew Roles amendment. Same Result-union shape as AssignResult/
// HireResult -- slot-full and Crafter-without-profession are normal
// business rejections a UI must handle gracefully, not caller errors.
export interface AssignShipRoleSucceeded {
  assigned: true;
  updatedCrewMember: CrewMember;
}

export interface AssignShipRoleRejected {
  assigned: false;
  reason: string;
}

export type AssignShipRoleResult = AssignShipRoleSucceeded | AssignShipRoleRejected;

File: unity/ProfitableCore/Schema/AssignShipRoleResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/assignShipRoleResult.ts's AssignShipRoleResult =
// AssignShipRoleSucceeded | AssignShipRoleRejected discriminated union.
public abstract class AssignShipRoleResult
{
    public abstract bool Assigned { get; }
}

public sealed class AssignShipRoleSucceeded : AssignShipRoleResult
{
    public override bool Assigned => true;
    public CrewMember UpdatedCrewMember { get; init; } = new();
}

public sealed class AssignShipRoleRejected : AssignShipRoleResult
{
    public override bool Assigned => false;
    public string Reason { get; init; } = string.Empty;
}
